### ISSUE TITLE
CAP_NET_ADMIN should permit tun devices to be placed into 'tap' mode

### ISSUE THREAD

**Issue by adamierymenko** [docker/docker#17532]
It's possible to get a tun device in a Docker container with "--device=/dev/net/tun", but it can't be used in "tap" mode with CAP_NET_ADMIN. In particular, this code (run inside the container) fails with only CAP_NET_ADMIN:

    ifr.ifr_flags = IFF_TAP | IFF_NO_PI;
    if (ioctl(_fd,TUNSETIFF,(void *)&ifr) < 0) {
        close(_fd);
        throw std::runtime_error("unable to configure TUN/TAP device for TAP operation");
    }

It looks like ioctl(fd,TUNSETIFF,...) is not permitted.

It seems to me that CAP_NET_ADMIN should permit the tun to be placed into tap mode, as this is a normal operation to perform on one's own tun device. I think this might be a Linux kernel bug, but I'm posting it here since I'm not sure where else to put it and I'm not sure whether it's Docker or the Linux kernel that configures the specific whitelists associated with these capabilities.

Right now to make the above work requires "--privileged", which basically strips away all protection and allows the container to do (almost) anything. This is workable for testing but is not desirable in a production environment.

**Comment by adamierymenko** [comment]
Docker environment info:

[root@linux-dev ~]# docker --version
Docker version 1.7.1, build 446ad9b/1.7.1
[root@linux-dev ~]# docker info
Containers: 0
Images: 20
Storage Driver: devicemapper
 Pool Name: docker-253:0-2229057-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 609.2 MB
 Data Space Total: 107.4 GB
 Data Space Available: 50.2 GB
 Metadata Space Used: 1.376 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.10.0-229.14.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 2
Total Memory: 1.794 GiB
Name: linux-dev
ID: SDXE:RBKP:MCUJ:42F5:63AH:IHPB:IUAB:2QYK:TJ2E:UBBJ:V2OI:R6FP
Username: adamierymenko
Registry: https://index.docker.io/v1/
[root@linux-dev ~]# uname -a
Linux linux-dev 3.10.0-229.14.1.el7.x86_64 #1 SMP Tue Sep 15 15:05:51 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Steps to reproduce:

1. Run a container with "--device=/dev/net/tun --cap-add=CAP_NET_ADMIN".

2. Try to run anything that puts a tun device in tap mode (OpenVPN in tap mode, ZeroTier One, etc.)

3. The above ioctl will fail.

**Comment by cpuguy83** [comment]
I'm not sure what you are expecting here.
Capabilities come from the kernel.
Also are you sure selinux isn't blocking this?

**Comment by abhilashgupta** [comment]
I got the following error: "ioctl(TUNSETIFF): Operation not permitted" when I tried to create tap interfaces on tun devices, with just "--device=/dev/net/tun" option, and no "--cap-add" option.
So, I tried adding the "--cap-add=CAP_NET_ADMIN" option, to run the docker image, and got the following error when I tried to run it: "docker: Error response from daemon: linux spec capabilities: Unknown capability to add: "CAP_CAP_NET_ADMIN""
So, I changed my --cap-add option to "--cap-add=NET_ADMIN" and it worked! I was able to successfully add tap interfaces.
System config:
Docker version:
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:42:09 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:42:09 2016
 OS/Arch:      linux/amd64

Docker info:
Containers: 19
 Running: 2
 Paused: 0
 Stopped: 17
Images: 12
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 67
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-32-generic
Operating System: Ubuntu precise (12.04.5 LTS)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 993.7 MiB
Name: abhilash-VirtualBox
ID: IMK5:ZAOZ:H4HN:Z6OD:2LSK:LHW3:6M6K:VAIU:OGIV:UCI2:NWIX:V5DT
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support

uname -a:
Linux 3.13.0-32-generic #57~precise1-Ubuntu SMP Tue Jul 15 03:51:20 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
